File: Binary Trees/max_path_sum.py
def max_path_sum(tree):
    _, max_sum = find_sum(tree)
    return max_sum

def find_sum(node):
    if not node:
        return (0,0)
    
    left_max_branch, left_max_path = find_sum(node.left)
    right_max_branch, right_max_path = find_sum(node.right)
    max_child_branch = max(left_max_branch, right_max_branch)

    value = node.value
    max_branch = max(max_child_branch + value, value)
    max_at_root = max(left_max_branch + value + right_max_branch, max_branch)
    max_path = max(left_max_path, right_max_path, max_at_root)

    return (max_branch, max_path)

File: Binary Trees/maxPathSum.js
const maxPathSum = tree => {
    const [_, maxSum] = [...findSums(tree)]
	return maxSum
}

const findSums = (node) => {
	if (node === null) return [0,0]
	
	const [leftMaxBranch, leftMaxPath] = findSums(node.left)
	const [rightMaxBranch, rightMaxPath] = findSums(node.right)
	const maxChildBranch = Math.max(leftMaxBranch, rightMaxBranch)
	
	const {value} = node
	const maxBranch = Math.max(maxChildBranch + value, value)
	const maxAtRoot = Math.max(leftMaxBranch + value + rightMaxBranch, maxBranch)
	const maxPath = Math.max(maxAtRoot, leftMaxPath, rightMaxPath)
	
	return [maxBranch, maxPath]
}